README.md
# python-docraptor

A wrapper for the Doc Raptor API.

# Example

    from docraptor import DocRaptor

    docraptor = DocRaptor()
    with open("test.pdf", "wb") as f:
        f.write(docraptor.create({
            'document_content': '<p>python-docraptor Test</p>', 
            'test': True
        }).content)

# Async Example

    import time
    from docraptor import DocRaptor

    docraptor = DocRaptor()

    resp = docraptor.create({
        'document_content': '<p>python-docraptor Async Test</p>', 
        'test': True, 
        'async': True 
    })
    status_id = resp['status_id']

    resp = docraptor.status(status_id)
    while resp['status'] != 'completed':
        time.sleep(3)
        resp = docraptor.status(status_id)
    
    with open("test_async.pdf", "wb") as f:
        f.write(docraptor.download(resp['download_key']).content)

# Build Status

[![Build Status](https://travis-ci.org/jkeyes/python-docraptor.png?branch=master)](https://travis-ci.org/jkeyes/python-docraptor)


HISTORY.rst
.. :changelog:

History
-------

1.2.3 (2019-01-09)
++++++++++++++++++

- Fix exception `.message` bug.

1.2.2 (2016-01-11)
++++++++++++++++++

- Increase default timeout to 90 seconds (@jamespaden)

1.2.1 (2013-08-12)
++++++++++++++++++

- Decode UTF-8 JSON responses

1.2 (2013-08-11)
++++++++++++++++++

- Python 3.3.2 compatibility


1.1 (2013-05-23)
++++++++++++++++++

- Added timeout to HTTP requests (@jesseh)


1.0 (2011-11-29)
++++++++++++++++++

- Created


test/tests.py
# Copyright (C) 2011 John Keyes
# http://jkeyes.mit-license.org/

import os
from os import path
from docraptor import DocRaptor
from docraptor import DocumentCreationFailure
from docraptor import DocumentStatusFailure
from docraptor import DocumentListingFailure
from docraptor import DocumentDownloadFailure
from docraptor import NoApiKeyProvidedError
from docraptor import NoContentError
import requests
from unittest import TestCase
from nose.tools import raises

TEST = "test.pdf"
TEST_NAME = "test_name.pdf"
FILES = [TEST, TEST_NAME]
DIRPATH = os.path.dirname(__file__)
FIXTURES = {}

def _clean():
    for fname in FILES:
        if path.exists(fname):
            os.unlink(fname)

class MockResponse(object):
    pass

def stub_http_response_with(filename, method=None, status=None):
    if status is None:
        status = 200
    try:
        # python 2
        content = open(os.path.join(DIRPATH, 'fixtures', filename), "r").read()
    except UnicodeDecodeError:
        # python 3
        content = open(os.path.join(DIRPATH, 'fixtures', filename), "r",
            encoding="latin-1").read()
    FIXTURES[filename] = content
    def stubbed(*args, **kwargs):
        resp = MockResponse()
        resp.status_code = status
        resp.content = content
        return resp
    setattr(requests, method, stubbed)

class DocRaptorApiKeyTest(TestCase):

    def setUp(self):
        os.environ.pop('DOCRAPTOR_API_KEY', None)
        self.test_key = "test key"

    @raises(NoApiKeyProvidedError)
    def test_api_keys(self):
        docraptor = DocRaptor()

    def test_env_api_key(self):
        os.environ['DOCRAPTOR_API_KEY'] = self.test_key
        docraptor = DocRaptor()
        assert docraptor.api_key == self.test_key

    def test_no_overwrite_env_api_key(self):
        os.environ['DOCRAPTOR_API_KEY'] = self.test_key
        docraptor = DocRaptor()

        os.environ['DOCRAPTOR_API_KEY'] = "blah"
        assert docraptor.api_key == self.test_key

    def test_param_api_keys(self):
        docraptor = DocRaptor(self.test_key)
        assert docraptor.api_key == self.test_key

class DocRaptorCreateTest(TestCase):
    def setUp(self):
        self.test_key = "test key"

    @raises(AttributeError)
    def test_bogus_arguments_bool(self):
        DocRaptor(self.test_key).create(True)

    @raises(ValueError)
    def test_bogus_arguments_none(self):
        DocRaptor(self.test_key).create(None)

    @raises(NoContentError)
    def test_no_content_empty(self):
        DocRaptor(self.test_key).create({})

    @raises(NoContentError)
    def test_no_content_attr(self):
        DocRaptor(self.test_key).create({'herped': 'the_derp'})

    @raises(NoContentError)
    def test_blank_content(self):
        DocRaptor(self.test_key).create({ 'content': ''})

    @raises(NoContentError)
    def test_blank_url(self):
        DocRaptor(self.test_key).create({'url': ''})
    
class DocRaptorDocumentContentTest(TestCase):
    def setUp(self):
        self.test_key = "test key"
        self.html_content = "<html><body>Hey</body></html>"

    def test_create_with_invalid_content(self):
        invalid_html = "<herp"
        stub_http_response_with("invalid_pdf", "post", 422)
        resp = DocRaptor(self.test_key).create({'document_content': invalid_html })
        assert FIXTURES["invalid_pdf"] == resp.content
        assert 422 == resp.status_code

    @raises(DocumentCreationFailure)
    def test_create_with_invalid_content_raises(self):
        invalid_html = "<herp"
        stub_http_response_with("invalid_pdf", "post", 422)
        resp = DocRaptor(self.test_key).create({'document_content': invalid_html, 'raise_exception_on_failure': True })

    def test_create_with_valid_content(self):
        stub_http_response_with("simple_pdf", "post")
        resp = DocRaptor(self.test_key).create({'document_content': self.html_content })
        assert FIXTURES["simple_pdf"] == resp.content
        assert 200 == resp.status_code

class DocRaptorListDocsTest(TestCase):
    def setUp(self):
        self.test_key = "test key"

    @raises(AttributeError)
    def test_bogus_arguments_bool(self):
        DocRaptor(self.test_key).list_docs(True)

    @raises(ValueError)
    def test_bogus_arguments_None(self):
        DocRaptor(self.test_key).list_docs(None)

    @raises(DocumentListingFailure)
    def test_invalid_list_docs_raises(self):
        stub_http_response_with("invalid_list_docs", "get", 400)
        resp = DocRaptor(self.test_key).list_docs({'raise_exception_on_failure': True})

    def test_list_docs(self):
        stub_http_response_with("simple_list_docs", "get")
        resp = DocRaptor(self.test_key).list_docs({})
        assert FIXTURES["simple_list_docs"] == resp.content

class DocRaptorStatusTest(TestCase):
    def setUp(self):
        self.test_key = "test key"

    def test_invalid_status(self):
        stub_http_response_with("invalid_status", "get", 403)
        resp = DocRaptor(self.test_key).status("test-id")
        assert FIXTURES["invalid_status"] == resp.content
        assert 403 == resp.status_code

    @raises(DocumentStatusFailure)
    def test_invalid_status_raises(self):
        stub_http_response_with("invalid_status", "get", 400)
        resp = DocRaptor(self.test_key).status("test-id", True)

    def atest_status(self):
        stub_http_response_with("simple_status", "get")
        resp = DocRaptor(self.test_key).status("test-id")
        assert "completed" == resp['status']

class DocRaptorDownloadTest(TestCase):
    def setUp(self):
        self.test_key = "test key"

    def test_invalid_download(self):
        stub_http_response_with("invalid_download", "get", 400)
        resp = DocRaptor(self.test_key).download("test-id")
        assert FIXTURES["invalid_download"] == resp.content
        assert 400 == resp.status_code

    @raises(DocumentDownloadFailure)
    def test_invalid_download_raises(self):
        stub_http_response_with("invalid_download", "get", 400)
        resp = DocRaptor(self.test_key).download("test-id", True)
        
    def test_download(self):
        stub_http_response_with("simple_download", "get")
        resp = DocRaptor(self.test_key).download("test-id")
        assert FIXTURES["simple_download"] == resp.content


tests/tests.py
# Copyright (C) 2011 John Keyes
# http://jkeyes.mit-license.org/

import os
from os import path
from docraptor import DocRaptor
from docraptor import DocumentCreationFailure
from docraptor import DocumentStatusFailure
from docraptor import DocumentListingFailure
from docraptor import DocumentDownloadFailure
from docraptor import NoApiKeyProvidedError
from docraptor import NoContentError
import requests
from unittest import TestCase
from nose.tools import raises

TEST = "test.pdf"
TEST_NAME = "test_name.pdf"
FILES = [TEST, TEST_NAME]
DIRPATH = os.path.dirname(__file__)
FIXTURES = {}

def _clean():
    for fname in FILES:
        if path.exists(fname):
            os.unlink(fname)

class MockResponse(object):
    pass

def stub_http_response_with(filename, method=None, status=None):
    if status is None:
        status = 200
    try:
        # python 2
        content = open(os.path.join(DIRPATH, 'fixtures', filename), "r").read()
    except UnicodeDecodeError:
        # python 3
        content = open(os.path.join(DIRPATH, 'fixtures', filename), "r",
            encoding="latin-1").read()
    FIXTURES[filename] = content
    def stubbed(*args, **kwargs):
        resp = MockResponse()
        resp.status_code = status
        resp.content = content
        return resp
    setattr(requests, method, stubbed)

class DocRaptorApiKeyTest(TestCase):

    def setUp(self):
        os.environ.pop('DOCRAPTOR_API_KEY', None)
        self.test_key = "test key"

    @raises(NoApiKeyProvidedError)
    def test_api_keys(self):
        docraptor = DocRaptor()

    def test_env_api_key(self):
        os.environ['DOCRAPTOR_API_KEY'] = self.test_key
        docraptor = DocRaptor()
        assert docraptor.api_key == self.test_key

    def test_no_overwrite_env_api_key(self):
        os.environ['DOCRAPTOR_API_KEY'] = self.test_key
        docraptor = DocRaptor()

        os.environ['DOCRAPTOR_API_KEY'] = "blah"
        assert docraptor.api_key == self.test_key

    def test_param_api_keys(self):
        docraptor = DocRaptor(self.test_key)
        assert docraptor.api_key == self.test_key

class DocRaptorCreateTest(TestCase):
    def setUp(self):
        self.test_key = "test key"

    @raises(AttributeError)
    def test_bogus_arguments_bool(self):
        DocRaptor(self.test_key).create(True)

    @raises(ValueError)
    def test_bogus_arguments_none(self):
        DocRaptor(self.test_key).create(None)

    @raises(NoContentError)
    def test_no_content_empty(self):
        DocRaptor(self.test_key).create({})

    @raises(NoContentError)
    def test_no_content_attr(self):
        DocRaptor(self.test_key).create({'herped': 'the_derp'})

    @raises(NoContentError)
    def test_blank_content(self):
        DocRaptor(self.test_key).create({ 'content': ''})

    @raises(NoContentError)
    def test_blank_url(self):
        DocRaptor(self.test_key).create({'url': ''})

    @raises(requests.exceptions.Timeout)
    def test_timeout(self):
        os.environ['DOCRAPTOR_TIMEOUT'] = '0.0001'
        resp = DocRaptor(self.test_key).create({'document_content': "<html><body>Hey</body></html>" })

    
class DocRaptorDocumentContentTest(TestCase):
    def setUp(self):
        self.test_key = "test key"
        self.html_content = "<html><body>Hey</body></html>"

    def test_create_with_invalid_content(self):
        invalid_html = "<herp"
        stub_http_response_with("invalid_pdf", "post", 422)
        resp = DocRaptor(self.test_key).create({'document_content': invalid_html })
        assert FIXTURES["invalid_pdf"] == resp.content
        assert 422 == resp.status_code

    @raises(DocumentCreationFailure)
    def test_create_with_invalid_content_raises(self):
        invalid_html = "<herp"
        stub_http_response_with("invalid_pdf", "post", 422)
        resp = DocRaptor(self.test_key).create({'document_content': invalid_html, 'raise_exception_on_failure': True })

    def test_create_with_valid_content(self):
        stub_http_response_with("simple_pdf", "post")
        resp = DocRaptor(self.test_key).create({'document_content': self.html_content })
        assert FIXTURES["simple_pdf"] == resp.content
        assert 200 == resp.status_code

class DocRaptorListDocsTest(TestCase):
    def setUp(self):
        self.test_key = "test key"

    @raises(AttributeError)
    def test_bogus_arguments_bool(self):
        DocRaptor(self.test_key).list_docs(True)

    @raises(ValueError)
    def test_bogus_arguments_None(self):
        DocRaptor(self.test_key).list_docs(None)

    @raises(DocumentListingFailure)
    def test_invalid_list_docs_raises(self):
        stub_http_response_with("invalid_list_docs", "get", 400)
        resp = DocRaptor(self.test_key).list_docs({'raise_exception_on_failure': True})

    def test_list_docs(self):
        stub_http_response_with("simple_list_docs", "get")
        resp = DocRaptor(self.test_key).list_docs({})
        assert FIXTURES["simple_list_docs"] == resp.content

class DocRaptorStatusTest(TestCase):
    def setUp(self):
        self.test_key = "test key"

    def test_invalid_status(self):
        stub_http_response_with("invalid_status", "get", 403)
        resp = DocRaptor(self.test_key).status("test-id")
        assert FIXTURES["invalid_status"] == resp.content
        assert 403 == resp.status_code

    @raises(DocumentStatusFailure)
    def test_invalid_status_raises(self):
        stub_http_response_with("invalid_status", "get", 400)
        resp = DocRaptor(self.test_key).status("test-id", True)

    def atest_status(self):
        stub_http_response_with("simple_status", "get")
        resp = DocRaptor(self.test_key).status("test-id")
        assert "completed" == resp['status']

class DocRaptorDownloadTest(TestCase):
    def setUp(self):
        self.test_key = "test key"

    def test_invalid_download(self):
        stub_http_response_with("invalid_download", "get", 400)
        resp = DocRaptor(self.test_key).download("test-id")
        assert FIXTURES["invalid_download"] == resp.content
        assert 400 == resp.status_code

    @raises(DocumentDownloadFailure)
    def test_invalid_download_raises(self):
        stub_http_response_with("invalid_download", "get", 400)
        resp = DocRaptor(self.test_key).download("test-id", True)
        
    def test_download(self):
        stub_http_response_with("simple_download", "get")
        resp = DocRaptor(self.test_key).download("test-id")
        assert FIXTURES["simple_download"] == resp.content


docraptor/tests.py
# Copyright (C) 2011 John Keyes
# http://jkeyes.mit-license.org/

import os
from os import path
from . import pdf
from . import xls
from unittest import TestCase

TEST = "test.pdf"
TEST_NAME = "test_name.pdf"
FILES = [TEST, TEST_NAME]

KEY = os.environ.get('DOCRAPTOR_KEY', None)

def _clean():
    for fname in FILES:
        if path.exists(fname):
            os.unlink(fname)

class PDFTest(TestCase):

    def setUp(self):
        _clean()

    def tearDown(self):
        _clean()

    def test_pdf(self):
        assert not path.exists(TEST)
        pdf(KEY,
            { 'doc[test]': 'true',
              'doc[name]': TEST,
              'doc[document_content]': "<html><body><h1>Test</h1></body></html>"})
        assert path.exists("test.pdf")

    def test_ssl(self):
        assert not path.exists(TEST)
        pdf(KEY,
            { 'doc[test]': 'true',
              'doc[name]': TEST,
              'doc[document_content]':"<html><body><h1>Test</h1></body></html>"},
            ssl=True)
        assert path.exists("test.pdf")

    def test_name(self):
        assert not path.exists(TEST)
        assert not path.exists(TEST_NAME)
        tnf = open("test_name.pdf", "w")
        pdf(KEY,
            { 'doc[test]': 'true',
              'doc[name]': "test.pdf",
              'doc[document_content]':"<html><body><h1>Test</h1></body></html>"},
            output_file=tnf,
            ssl=True)
        assert not path.exists(TEST)
        assert path.exists(TEST_NAME)

    def test_content_or_url_required(self):
        try:
            pdf(KEY,
                'doc[test]': "true",
                'doc[name]': "test"
                )
        except DocRaptorError, e:
            assert e.message == "Must provide a document url or document content"


docraptor/__init__.py
# Copyright (C) 2011 John Keyes
# http://jkeyes.mit-license.org/

"""
Simple wrapper for Doc Raptor API.
"""
import json
import os
import re
import requests


ENV = os.environ
API_KEY = "DOCRAPTOR_API_KEY"
URL = "DOCRAPTOR_URL"
DEFAULT_TIMEOUT = 90  # seconds
TIMEOUT = "DOCRAPTOR_TIMEOUT"

# endpoint URLs
HTTPS_URL = 'https://docraptor.com/'


class NoApiKeyProvidedError(RuntimeError):
    pass


class NoContentError(KeyError):
    pass


class DocRaptorRequestException(Exception):
    def __init__(self, message, status_code):
        super(DocRaptorRequestException, self).__init__(message)
        self.message = message
        self.status_code = status_code

    def __str__(self):
        return "%s\nHTTP Status: %s\nReturned: %s" % (
            self.__class__.__name__, self.status_code, self.message)


class DocumentCreationFailure(DocRaptorRequestException):
    pass


class DocumentListingFailure(DocRaptorRequestException):
    pass


class DocumentStatusFailure(DocRaptorRequestException):
    pass


class DocumentDownloadFailure(DocRaptorRequestException):
    pass


class DocRaptor(object):

    def __init__(self, api_key=None):
        self.api_key = ENV.get(API_KEY) if api_key is None else api_key
        if not self.api_key:
            raise NoApiKeyProvidedError("No API key provided")
        self._url = ENV.get(URL, HTTPS_URL)
        self._timeout = float(ENV.get(TIMEOUT, DEFAULT_TIMEOUT))

    def create(self, options=None):
        if options is None:
            raise ValueError("Please pass in an options dict")

        if not _has_content(options):
            raise NoContentError(
                "must supply 'document_content' or 'document_url'")

        default_options = {
            'name': 'default',
            'document_type': 'pdf',
            'test': False,
            'async': False,
            'raise_exception_on_failure': False
        }
        options = dict(list(default_options.items()) + list(options.items()))
        raise_exception_on_failure = options.pop('raise_exception_on_failure')
        query = {'user_credentials': self.api_key}
        if options['async']:
            query['output'] = 'json'
        doc_options = _format_keys({'doc': options})

        resp = requests.post(
            '%sdocs' % (self._url), data=doc_options, params=query,
            timeout=self._timeout)

        if raise_exception_on_failure and resp.status_code != 200:
            raise DocumentCreationFailure(resp.content, resp.status_code)

        if options['async']:
            return json.loads(resp.content.decode("utf-8"))
        else:
            return resp

    def list_docs(self, options=None):
        if options is None:
            raise ValueError("Please pass in an options dict")

        default_options = {
            'page': 1,
            'per_page': 100,
            'raise_exception_on_failure': False,
            'user_credentials': self.api_key
        }
        options = dict(list(default_options.items()) + list(options.items()))
        raise_exception_on_failure = options.pop('raise_exception_on_failure')

        resp = requests.get(
            '%sdocs' % (self._url), params=options, timeout=self._timeout)

        if raise_exception_on_failure and resp.status_code != 200:
            raise DocumentListingFailure(resp.content, resp.status_code)
        return resp

    def status(self, status_id, raise_exception_on_failure=False):
        query = {
            'output': 'json',
            'user_credentials': self.api_key
        }

        resp = requests.get(
            '%sstatus/%s' % (self._url, status_id), params=query,
            timeout=self._timeout)

        if raise_exception_on_failure and resp.status_code != 200:
            raise DocumentStatusFailure(resp.content, resp.status_code)

        if resp.status_code == 200:
            as_json = json.loads(resp.content.decode("utf-8"))
            if as_json['status'] == 'completed':
                match = re.match(
                    '.*?\/download\/(.+)', as_json['download_url'])
                as_json['download_key'] = match.groups()[0]
            return as_json
        return resp

    def download(self, download_key, raise_exception_on_failure=False):
        query = {
            'output': 'json',
            'user_credentials': self.api_key
        }
        resp = requests.get(
            '%sdownload/%s' % (self._url, download_key), params=query,
            timeout=self._timeout)

        if raise_exception_on_failure and resp.status_code != 200:
            raise DocumentDownloadFailure(resp.content, resp.status_code)
        return resp


def _has_content(options):
    content = options.get('document_content') or options.get('document_url')
    return bool(content)


def _format_keys(options, result=None, parent_key=None):
    if result is None:
        result = {}
    for k, v in list(options.items()):
        if parent_key:
            key = '%s[%s]' % (parent_key, k)
        else:
            key = k
        if isinstance(v, dict):
            _format_keys(v, result, key)
        elif v is False:
            result[key] = 'false'
        elif v is True:
            result[key] = 'true'
        else:
            result[key] = v
    return result
